File: examples/language-impl-comparison/elixir/markdown_parser.exs
defmodule MarkdownParser do
  @moduledoc """
  Markdown風軽量マークアップパーサー - Elixir実装

  Unicode処理の注意点：
  - ElixirのStringはUTF-8バイナリで、graphemeクラスター単位の操作が可能
  - String.length/1 はgrapheme数を返す
  - byte_size/1 はバイト数を返す
  - String.codepoints/1 でコードポイント単位の操作も可能
  - Remlの3層モデル（Byte/Char/Grapheme）に近い柔軟性を持つ
  """

  @type inline ::
          {:text, String.t()}
          | {:strong, [inline()]}
          | {:emphasis, [inline()]}
          | {:code, String.t()}
          | {:link, [inline()], String.t()}
          | :line_break

  @type block ::
          {:heading, non_neg_integer(), [inline()]}
          | {:paragraph, [inline()]}
          | {:unordered_list, [[inline()]]}
          | {:ordered_list, [[inline()]]}
          | {:code_block, String.t() | nil, String.t()}
          | :horizontal_rule

  @type document :: [block()]

  @type parse_state :: %{
          input: String.t(),
          position: non_neg_integer()
        }

  @type parse_result(t) :: {:ok, t, parse_state()} | {:error, String.t()}

  @spec parse(String.t()) :: {:ok, document()} | {:error, String.t()}
  def parse(input) do
    initial_state = %{input: input, position: 0}
    parse_document(initial_state, [])
  end

  defp parse_document(state, blocks) do
    case parse_block(state) do
      {:ok, block, new_state} ->
        parse_document(new_state, [block | blocks])

      {:error, "EOF"} ->
        {:ok, Enum.reverse(blocks)}

      {:error, reason} ->
        {:error, reason}
    end
  end

  defp parse_block(state) do
    state = skip_blank_lines(state)

    if eof?(state) do
      {:error, "EOF"}
    else
      state = skip_hspace(state)

      case peek_char(state) do
        {:ok, ?#} -> parse_heading(state)
        {:ok, ?`} ->
          case match_string(state, "```") do
            {:ok, _} -> parse_code_block(state)
            :error -> parse_paragraph(state)
          end

        {:ok, c} when c in [?-, ?*, ?_] ->
          case parse_horizontal_rule(state) do
            {:ok, _block, _new_state} = result -> result
            {:error, _} -> parse_unordered_list(state)
          end

        {:ok, _} ->
          parse_paragraph(state)

        :error ->
          {:error, "EOF"}
      end
    end
  end

  defp parse_heading(state) do
    state = skip_hspace(state)
    {level, state} = count_hashes(state, 0)

    if level == 0 or level > 6 do
      {:error, "見出しレベルは1-6の範囲内である必要があります"}
    else
      state = skip_hspace(state)
      {text, state} = read_until_eol(state)
      state = consume_newline(state)
      inline = [{:text, String.trim(text)}]
      {:ok, {:heading, level, inline}, state}
    end
  end

  defp count_hashes(state, n) do
    case peek_char(state) do
      {:ok, ?#} -> count_hashes(advance_char(state), n + 1)
      _ -> {n, state}
    end
  end

  defp parse_horizontal_rule(state) do
    state = skip_hspace(state)
    {text, state} = read_until_eol(state)
    state = consume_newline(state)

    trimmed = String.trim(text)

    is_rule =
      (String.graphemes(trimmed) |> Enum.all?(&(&1 == "-")) and String.length(trimmed) >= 3) or
        (String.graphemes(trimmed) |> Enum.all?(&(&1 == "*")) and String.length(trimmed) >= 3) or
        (String.graphemes(trimmed) |> Enum.all?(&(&1 == "_")) and String.length(trimmed) >= 3)

    if is_rule do
      {:ok, :horizontal_rule, state}
    else
      {:error, "水平線として認識できません"}
    end
  end

  defp parse_code_block(state) do
    case match_string(state, "```") do
      :error ->
        {:error, "コードブロック開始が見つかりません"}

      {:ok, state} ->
        {lang_line, state} = read_until_eol(state)
        state = consume_newline(state)

        lang =
          case String.trim(lang_line) do
            "" -> nil
            trimmed -> trimmed
          end

        {code_lines, state} = read_code_lines(state, [])
        state = consume_newline(state)

        code = Enum.join(code_lines, "\n")
        {:ok, {:code_block, lang, code}, state}
    end
  end

  defp read_code_lines(state, acc) do
    case match_string(state, "```") do
      {:ok, state} ->
        {Enum.reverse(acc), state}

      :error ->
        if eof?(state) do
          {Enum.reverse(acc), state}
        else
          {line, state} = read_until_eol(state)
          state = consume_newline(state)
          read_code_lines(state, [line | acc])
        end
    end
  end

  defp parse_unordered_list(state) do
    {items, state} = parse_list_items(state, [])

    if Enum.empty?(items) do
      {:error, "リスト項目が見つかりません"}
    else
      {:ok, {:unordered_list, Enum.reverse(items)}, state}
    end
  end

  defp parse_list_items(state, acc) do
    state = skip_hspace(state)

    case peek_char(state) do
      {:ok, c} when c in [?-, ?*] ->
        state = advance_char(state)
        state = skip_hspace(state)
        {text, state} = read_until_eol(state)
        state = consume_newline(state)
        inline = [{:text, String.trim(text)}]
        parse_list_items(state, [inline | acc])

      _ ->
        {acc, state}
    end
  end

  defp parse_paragraph(state) do
    {lines, state} = read_paragraph_lines(state, [])
    text = lines |> Enum.join(" ") |> String.trim()
    inline = [{:text, text}]
    {:ok, {:paragraph, inline}, state}
  end

  defp read_paragraph_lines(state, acc) do
    if eof?(state) do
      {Enum.reverse(acc), state}
    else
      case peek_char(state) do
        {:ok, ?\n} ->
          state = advance_char(state)

          case peek_char(state) do
            {:ok, ?\n} -> {Enum.reverse(acc), state}
            _ -> read_paragraph_lines(state, ["" | acc])
          end

        {:ok, _} ->
          {line, state} = read_until_eol(state)
          state = consume_newline(state)
          read_paragraph_lines(state, [line | acc])

        :error ->
          {Enum.reverse(acc), state}
      end
    end
  end

  # パーサーユーティリティ

  defp peek_char(%{input: input, position: pos}) do
    case String.at(input, pos) do
      nil -> :error
      char -> {:ok, String.to_charlist(char) |> hd()}
    end
  end

  defp advance_char(%{position: pos} = state) do
    %{state | position: pos + 1}
  end

  defp match_string(state, target) do
    %{input: input, position: pos} = state
    remaining = String.slice(input, pos..-1//1)

    if String.starts_with?(remaining, target) do
      new_pos = pos + String.length(target)
      {:ok, %{state | position: new_pos}}
    else
      :error
    end
  end

  defp skip_hspace(state) do
    case peek_char(state) do
      {:ok, c} when c in [?\s, ?\t] ->
        skip_hspace(advance_char(state))

      _ ->
        state
    end
  end

  defp skip_blank_lines(state) do
    case peek_char(state) do
      {:ok, ?\n} -> skip_blank_lines(advance_char(state))
      _ -> state
    end
  end

  defp read_until_eol(%{input: input, position: pos} = state) do
    remaining = String.slice(input, pos..-1//1)

    case String.split(remaining, "\n", parts: 2) do
      [line, _rest] ->
        {line, %{state | position: pos + String.length(line)}}

      [line] ->
        {line, %{state | position: pos + String.length(line)}}
    end
  end

  defp consume_newline(state) do
    case peek_char(state) do
      {:ok, ?\n} -> advance_char(state)
      _ -> state
    end
  end

  defp eof?(%{input: input, position: pos}) do
    pos >= String.length(input)
  end

  # レンダリング

  @spec render_to_string(document()) :: String.t()
  def render_to_string(doc) do
    doc |> Enum.map(&render_block/1) |> Enum.join()
  end

  defp render_inline(inlines) do
    inlines
    |> Enum.map(fn
      {:text, text} -> text
      {:strong, inner} -> "**#{render_inline(inner)}**"
      {:emphasis, inner} -> "*#{render_inline(inner)}*"
      {:code, code} -> "`#{code}`"
      {:link, text, url} -> "[#{render_inline(text)}](#{url})"
      :line_break -> "\n"
    end)
    |> Enum.join()
  end

  defp render_block(block) do
    case block do
      {:heading, level, inline} ->
        prefix = String.duplicate("#", level)
        "#{prefix} #{render_inline(inline)}\n\n"

      {:paragraph, inline} ->
        "#{render_inline(inline)}\n\n"

      {:unordered_list, items} ->
        items_str =
          items
          |> Enum.map(fn item -> "- #{render_inline(item)}\n" end)
          |> Enum.join()

        "#{items_str}\n"

      {:ordered_list, items} ->
        items_str =
          items
          |> Enum.with_index(1)
          |> Enum.map(fn {item, i} -> "#{i}. #{render_inline(item)}\n" end)
          |> Enum.join()

        "#{items_str}\n"

      {:code_block, lang, code} ->
        lang_str = if lang, do: lang, else: ""
        "```#{lang_str}\n#{code}\n```\n\n"

      :horizontal_rule ->
        "---\n\n"
    end
  end
end

# Unicode 3層モデル比較：
#
# Elixirでは String はUTF-8バイナリで、標準で3つのレベルの操作が可能：
#
# - byte_size(str) → バイト数
# - String.length(str) → Grapheme（書記素クラスター）数
# - String.codepoints(str) → コードポイント単位のリスト
#
# 例：
# str = "🇯🇵"  # 国旗絵文字（2つのコードポイント、1つのgrapheme）
# byte_size(str) # => 8
# String.length(str) # => 1 (grapheme)
# String.codepoints(str) # => ["🇯", "🇵"]
#
# ElixirはRemlの3層モデルに近い明示性を持ち、
# デフォルトでgrapheme単位の操作が可能なため、絵文字や結合文字の扱いが自然。

File: examples/language-impl-comparison/fsharp/markdown_parser.fs
// Markdown風軽量マークアップパーサー - F#実装
//
// Unicode処理の注意点：
// - F#のstringは.NETのString（UTF-16ベース）
// - String.length はUTF-16コードユニット数を返す（サロゲートペアに注意）
// - .NET Core 3.0以降はRune型でUnicodeコードポイント単位の操作が可能
// - Grapheme（書記素クラスター）処理にはSystem.Globalization.StringInfoが必要
// - Remlの3層モデルと比較すると、F#も明示的な区別が必要

module MarkdownParser

open System

/// Markdown AST のインライン要素
type Inline =
  | Text of string
  | Strong of Inline list
  | Emphasis of Inline list
  | Code of string
  | Link of text: Inline list * url: string
  | LineBreak

/// Markdown AST のブロック要素
type Block =
  | Heading of level: int * inline: Inline list
  | Paragraph of inline: Inline list
  | UnorderedList of items: Inline list list
  | OrderedList of items: Inline list list
  | CodeBlock of lang: string option * code: string
  | HorizontalRule

type Document = Block list

/// パーサー状態
type ParseState =
  { Input: string
    Position: int }

type ParseResult<'T> = Result<'T * ParseState, string>

/// 現在位置の1文字を取得
let peekChar (state: ParseState) : char option =
  if state.Position >= state.Input.Length then
    None
  else
    Some state.Input.[state.Position]

/// 1文字を消費して進める
let advanceChar (state: ParseState) : ParseState =
  { state with Position = state.Position + 1 }

/// 固定文字列をマッチ
let matchString (target: string) (state: ParseState) : ParseState option =
  let remaining = state.Input.Substring(state.Position)
  if remaining.StartsWith(target) then
    Some { state with Position = state.Position + target.Length }
  else
    None

/// 水平空白をスキップ
let rec skipHSpace (state: ParseState) : ParseState =
  match peekChar state with
  | Some c when c = ' ' || c = '\t' ->
      skipHSpace (advanceChar state)
  | _ -> state

/// 空行をスキップ
let rec skipBlankLines (state: ParseState) : ParseState =
  match peekChar state with
  | Some '\n' -> skipBlankLines (advanceChar state)
  | _ -> state

/// 行末まで読む
let readUntilEol (state: ParseState) : string * ParseState =
  let rec loop (pos: int) =
    if pos >= state.Input.Length then
      pos
    elif state.Input.[pos] = '\n' then
      pos
    else
      loop (pos + 1)

  let endPos = loop state.Position
  let text = state.Input.Substring(state.Position, endPos - state.Position)
  (text, { state with Position = endPos })

/// 改行を消費
let consumeNewline (state: ParseState) : ParseState =
  match peekChar state with
  | Some '\n' -> advanceChar state
  | _ -> state

/// EOFチェック
let isEof (state: ParseState) : bool =
  state.Position >= state.Input.Length

/// 見出し行のパース（`# Heading` 形式）
let parseHeading (state: ParseState) : ParseResult<Block> =
  let state1 = skipHSpace state

  // `#` の連続をカウント
  let rec countHashes (s: ParseState) (n: int) : int * ParseState =
    match peekChar s with
    | Some '#' -> countHashes (advanceChar s) (n + 1)
    | _ -> (n, s)

  let (level, state2) = countHashes state1 0

  if level = 0 || level > 6 then
    Error "見出しレベルは1-6の範囲内である必要があります"
  else
    let state3 = skipHSpace state2
    let (text, state4) = readUntilEol state3
    let state5 = consumeNewline state4
    let inline = [Text (text.Trim())]
    Ok (Heading (level, inline), state5)

/// 水平線のパース（`---`, `***`, `___`）
let parseHorizontalRule (state: ParseState) : ParseResult<Block> =
  let state1 = skipHSpace state
  let (text, state2) = readUntilEol state1
  let state3 = consumeNewline state2

  let trimmed = text.Trim()
  let isRule =
    (trimmed.ToCharArray() |> Array.forall ((=) '-') && trimmed.Length >= 3) ||
    (trimmed.ToCharArray() |> Array.forall ((=) '*') && trimmed.Length >= 3) ||
    (trimmed.ToCharArray() |> Array.forall ((=) '_') && trimmed.Length >= 3)

  if isRule then
    Ok (HorizontalRule, state3)
  else
    Error "水平線として認識できません"

/// コードブロックのパース（```言語名）
let parseCodeBlock (state: ParseState) : ParseResult<Block> =
  match matchString "```" state with
  | None -> Error "コードブロック開始が見つかりません"
  | Some state1 ->
      let (langLine, state2) = readUntilEol state1
      let state3 = consumeNewline state2

      let lang =
        let trimmed = langLine.Trim()
        if String.IsNullOrEmpty(trimmed) then None else Some trimmed

      // コードブロック内容を ```閉じまで読む
      let rec readCodeLines (s: ParseState) (lines: string list) : string list * ParseState =
        match matchString "```" s with
        | Some endState -> (List.rev lines, endState)
        | None ->
            if isEof s then
              (List.rev lines, s)
            else
              let (line, s2) = readUntilEol s
              let s3 = consumeNewline s2
              readCodeLines s3 (line :: lines)

      let (codeLines, state4) = readCodeLines state3 []
      let state5 = consumeNewline state4
      let code = String.Join("\n", codeLines)
      Ok (CodeBlock (lang, code), state5)

/// リスト項目のパース（簡易版：`-` または `*`）
let parseUnorderedList (state: ParseState) : ParseResult<Block> =
  let rec parseItems (s: ParseState) (items: Inline list list) : Inline list list * ParseState =
    let s1 = skipHSpace s
    match peekChar s1 with
    | Some c when c = '-' || c = '*' ->
        let s2 = advanceChar s1
        let s3 = skipHSpace s2
        let (text, s4) = readUntilEol s3
        let s5 = consumeNewline s4
        let inline = [Text (text.Trim())]
        parseItems s5 (items @ [inline])
    | _ -> (items, s)

  let (items, stateEnd) = parseItems state []

  if List.isEmpty items then
    Error "リスト項目が見つかりません"
  else
    Ok (UnorderedList items, stateEnd)

/// 段落のパース（簡易版：空行まで）
let parseParagraph (state: ParseState) : ParseResult<Block> =
  let rec readLines (s: ParseState) (lines: string list) : string list * ParseState =
    if isEof s then
      (List.rev lines, s)
    else
      match peekChar s with
      | Some '\n' ->
          let s1 = advanceChar s
          match peekChar s1 with
          | Some '\n' -> (List.rev lines, s1) // 空行で段落終了
          | _ -> readLines s1 ("" :: lines)
      | Some _ ->
          let (line, s2) = readUntilEol s
          let s3 = consumeNewline s2
          readLines s3 (line :: lines)
      | None -> (List.rev lines, s)

  let (lines, stateEnd) = readLines state []
  let text = String.Join(" ", lines).Trim()
  let inline = [Text text]
  Ok (Paragraph inline, stateEnd)

/// ブロック要素のパース（優先順位付き試行）
let rec parseBlock (state: ParseState) : ParseResult<Block> =
  let state1 = skipBlankLines state

  if isEof state1 then
    Error "EOF"
  else
    let state2 = skipHSpace state1

    match peekChar state2 with
    | Some '#' -> parseHeading state2
    | Some '`' ->
        match matchString "```" state2 with
        | Some _ -> parseCodeBlock state2
        | None -> parseParagraph state2
    | Some c when c = '-' || c = '*' || c = '_' ->
        match parseHorizontalRule state2 with
        | Ok result -> Ok result
        | Error _ -> parseUnorderedList state2
    | Some _ -> parseParagraph state2
    | None -> Error "EOF"

/// ドキュメント全体のパース
let rec parseDocument (state: ParseState) (blocks: Block list) : Result<Document, string> =
  match parseBlock state with
  | Ok (block, newState) ->
      parseDocument newState (blocks @ [block])
  | Error "EOF" -> Ok blocks
  | Error msg -> Error msg

/// パブリックAPI：文字列からドキュメントをパース
let parse (input: string) : Result<Document, string> =
  let initialState = { Input = input; Position = 0 }
  parseDocument initialState []

/// 簡易的なレンダリング（検証用）
let rec renderInline (inlines: Inline list) : string =
  inlines
  |> List.map (fun i ->
      match i with
      | Text s -> s
      | Strong inner -> "**" + renderInline inner + "**"
      | Emphasis inner -> "*" + renderInline inner + "*"
      | Code s -> "`" + s + "`"
      | Link (text, url) -> "[" + renderInline text + "](" + url + ")"
      | LineBreak -> "\n"
  )
  |> String.concat ""

let renderBlock (block: Block) : string =
  match block with
  | Heading (level, inline) ->
      let prefix = String.replicate level "#"
      prefix + " " + renderInline inline + "\n\n"
  | Paragraph inline ->
      renderInline inline + "\n\n"
  | UnorderedList items ->
      let itemsStr =
        items
        |> List.map (fun item -> "- " + renderInline item + "\n")
        |> String.concat ""
      itemsStr + "\n"
  | OrderedList items ->
      let itemsStr =
        items
        |> List.mapi (fun i item -> $"{i + 1}. " + renderInline item + "\n")
        |> String.concat ""
      itemsStr + "\n"
  | CodeBlock (lang, code) ->
      let langStr = Option.defaultValue "" lang
      "```" + langStr + "\n" + code + "\n```\n\n"
  | HorizontalRule ->
      "---\n\n"

let renderToString (doc: Document) : string =
  doc
  |> List.map renderBlock
  |> String.concat ""

// Unicode 3層モデル比較：
//
// F#の string は .NET の System.String（UTF-16ベース）なので：
// - String.length はUTF-16コードユニット数を返す
// - サロゲートペア（例：絵文字「😀」）は2カウントされる
// - .NET Core 3.0以降は System.Text.Rune 型でコードポイント単位の操作が可能
// - Grapheme処理には System.Globalization.StringInfo が必要
//
// 例：
// let str = "🇯🇵"  // 国旗絵文字（2つのコードポイント、1つのgrapheme）
// str.Length  // => 4 (UTF-16コードユニット数)
// StringInfo(str).LengthInTextElements  // => 1 (grapheme数)
//
// Remlの3層モデル（Byte/Char/Grapheme）と比較すると、
// F#も明示的な区別が必要で、標準APIだけでは絵文字や結合文字の扱いが複雑。